core/src/main/scala/org/bitcoins/core/crypto/MnemonicCode.scala

```diff
@@ -0,0 +1,247 @@
+      ): Option[Int] = {
+        val mid = start + (end - start) / 2
+        if (start > end) None // can't be found
+        else if (wordlist(mid) == word) Some(mid) // found
+        else if (wordlist(mid) > word)
+          search(start, mid - 1) // narrow the field
+        else search(mid + 1, end)
+      }
+
+      val indexOpt = search()
+
+      val index = indexOpt.getOrElse(
+        throw new IllegalArgumentException(s"Couldn't find $word in word list!")
+      )
+
+      BitVector.fromInt(index, size = MnemonicCode.BIT_GROUP_LENGTH)
+
+    }
+
+    val bits: BitVector = words
+      .map(curr => bitsForWord(curr))
+      .fold(BitVector.empty)((accum, curr) => accum ++ curr)
+
+    val codeInfo = MnemonicCode.getMnemonicCodeInfo(words)
+
+    bits.take(codeInfo.entropyAndChecksumBits)
+  }
+}
+
+/**
+  * @see [[https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki BIP39]]
+  */
+object MnemonicCode extends Factory[MnemonicCode] {
+
+  private[crypto] val BIT_GROUP_LENGTH = 11
+
+  private case class MnemonicCodeImpl(words: Vector[String])
+      extends MnemonicCode
+
+  /**
+    * Creates a mnemonic code from the provided words
+    */
+  def fromWords(words: Vector[String]): MnemonicCode = MnemonicCodeImpl(words)
+
+  /**
+    * Creates a mnemonic code from the provided bytes
+    */
+  override def fromBytes(bytes: ByteVector): MnemonicCode = {
+    val words = bytes.toHex.split(" ").toVector
+    fromWords(words)
+  }
+
+  /**
+    * Generates a mnemonic code from the given entropy bytes
+    */
+  def fromEntropy(entropy: ByteVector): MnemonicCode =
+    fromEntropy(entropy.bits)
+
+  /**
+    * Generates a mnemonic code from the given entropy bits
+    */
+  def fromEntropy(entropy: BitVector): MnemonicCode = {
+    val MIN_BITS = 128
+    require(
+      entropy.length >= MIN_BITS,
+      // $COVERAGE-OFF$
+      s"Entropy must be at least $MIN_BITS bits"
+    )
+    // $COVERAGE-ON$
+
+    val MAX_BITS = 256
+    require(
+      entropy.length <= MAX_BITS,
+      // $COVERAGE-OFF$
+      s"Entropy cannot be longer than $MAX_BITS bits"
+    )
+    // $COVERAGE-ON$
+
+    val ENTROPY_MULTIPLE = 32
+    require(
+      entropy.length % ENTROPY_MULTIPLE == 0,
+      // $COVERAGE-OFF$
+      s"Entropy must be a multiple of $ENTROPY_MULTIPLE bits!"
+    )
+    // $COVERAGE-ON$
+
+    val ENTROPY_CHECKSUM_DIVISOR = 32
+    val checksumLength = entropy.length / ENTROPY_CHECKSUM_DIVISOR
+
+    val hashedEntropy = CryptoUtil.sha256(entropy)
+    val hashedEntropyBits = hashedEntropy.bytes.toBitVector
+    val checkSum = hashedEntropyBits.take(checksumLength)
+    val entropyWithChecksum = entropy ++ checkSum
+
+    val bitGroups = entropyWithChecksum.grouped(BIT_GROUP_LENGTH)
+    val words = bitGroups.map { group =>
+      val index = group.toInt(signed = false)
+      ENGLISH_WORDS(index)
+    }.toVector
+
+    MnemonicCodeImpl(words)
+  }
+
+  private lazy val randomGenerator: SecureRandom = new SecureRandom
```

Comment:
I believe it is ok to cache this from reading the javadoc on `SecureRandom`. But I don't really like the idea of this sitting in memory for a long amount of time. In practice, it shouldn't hurt to make this a `def` no? I would be even more in favor or just moving it into `getEntropy` so it is in the smallest scope possible.

Reply:
I want our code that is generating entropy to have the smallest scope possible to make it harder for people that are modifying the code later to do anything weird with the PRNG

Reply:
Moved it into `getEntropy`